pkg/allocator/cache.go

```diff
@@ -159,17 +164,49 @@ func (c *cache) OnDelete(id idpool.ID, key AllocatorKey) {
 	c.onDeleteLocked(id, key)
 }
 
+func syncControllerName(id idpool.ID) string {
+	return fmt.Sprintf("sync-k8s-identity-%s", id)
+}
+
 // onDeleteLocked must be called while holding c.Mutex for writing
 func (c *cache) onDeleteLocked(id idpool.ID, key AllocatorKey) {
 	a := c.allocator
 	if a.enableMasterKeyProtection {
 		if value := a.localKeys.lookupID(id); value != nil {
-			ctx, cancel := context.WithTimeout(context.TODO(), backendOpTimeout)
-			defer cancel()
-			err := a.backend.UpdateKey(ctx, id, value, true)
-			if err != nil {
-				log.WithError(err).Errorf("OnDelete MasterKeyProtection update for key %q", id)
-			}
+			c.controllers.UpdateController(syncControllerName(id), controller.ControllerParams{
```

Comment:
Makes sense, I added a stop/drain call on stop following the watcher stopping: https://github.com/cilium/cilium/pull/28912/files#diff-a53523a55dfc263e2ba4590a90e6c21ecfeec344c7cda3011a81f1f993f78f6eR253

This would let any remaining dofuncs complete before closing down.